### ISSUE TITLE
Pre-puller can't pull images in v0.6

### ISSUE THREAD

**Issue by GerbenWelter** [jupyterhub/zero-to-jupyterhub-k8s#477]
As discussed on gitter:

Pre-puller fails to pull in required docker images or recognise the images are already available. It only keeps logging the following over and over again:

```
 2018/02/02 09:59:09 0 of 2 images pulled in node worker1.company.com
```

This happens either on an install or upgrade from v0.5 using Helm 2.7.2 and Kubernetes 1.8.4. It also happens on a test-cluster at home using Helm 2.8.0 and Kubernetes 1.9.2.

## Contents of `config.yaml`
```
proxy:
  secretToken: "random_hex_string"
  service:
    type: NodePort
```

The NodePort change is because this is an on-premise test-cluster.

And as requested by @yuvipanda the output (sanitized) of 'kubectl get nodes -o yaml':

```
apiVersion: v1
items:
- apiVersion: v1
  kind: Node
  metadata:
    annotations:
      flannel.alpha.coreos.com/backend-data: '{"VtepMAC":"5e:6b:ac:33:61:d5"}'
      flannel.alpha.coreos.com/backend-type: vxlan
      flannel.alpha.coreos.com/kube-subnet-manager: "true"
      flannel.alpha.coreos.com/public-ip: 10.1.1.1
      node.alpha.kubernetes.io/ttl: "0"
      volumes.kubernetes.io/controller-managed-attach-detach: "true"
    creationTimestamp: 2018-01-16T15:31:26Z
    labels:
      beta.kubernetes.io/arch: amd64
      beta.kubernetes.io/os: linux
      kubernetes.io/hostname: master.company.com
      node-role.kubernetes.io/master: ""
    name: master.company.com
    namespace: ""
    resourceVersion: "2355345"
    selfLink: /api/v1/nodes/master.company.com
    uid: 50e2a2d1-fad2-11e7-a5bb-0050568db8e3
  spec:
    externalID: master.company.com
    podCIDR: 10.244.0.0/24
    taints:
    - effect: NoSchedule
      key: node-role.kubernetes.io/master
      timeAdded: null
  status:
    addresses:
    - address: 10.1.1.1
      type: InternalIP
    - address: master.company.com
      type: Hostname
    allocatable:
      cpu: "2"
      memory: 3761852Ki
      pods: "110"
    capacity:
      cpu: "2"
      memory: 3864252Ki
      pods: "110"
    conditions:
    - lastHeartbeatTime: 2018-02-05T09:21:19Z
      lastTransitionTime: 2018-01-16T15:31:22Z
      message: kubelet has sufficient disk space available
      reason: KubeletHasSufficientDisk
      status: "False"
      type: OutOfDisk
    - lastHeartbeatTime: 2018-02-05T09:21:19Z
      lastTransitionTime: 2018-01-16T15:31:22Z
      message: kubelet has sufficient memory available
      reason: KubeletHasSufficientMemory
      status: "False"
      type: MemoryPressure
    - lastHeartbeatTime: 2018-02-05T09:21:19Z
      lastTransitionTime: 2018-01-16T15:31:22Z
      message: kubelet has no disk pressure
      reason: KubeletHasNoDiskPressure
      status: "False"
      type: DiskPressure
    - lastHeartbeatTime: 2018-02-05T09:21:19Z
      lastTransitionTime: 2018-01-16T15:31:56Z
      message: kubelet is posting ready status
      reason: KubeletReady
      status: "True"
      type: Ready
    daemonEndpoints:
      kubeletEndpoint:
        Port: 10250
    images:
    - names:
      - gcr.io/google_containers/kube-apiserver-amd64@sha256:47377f5df3ea25cb677fe081be78dce9ffe4a58dcaaa4191a0a9e8295cfb848b
      - gcr.io/google_containers/kube-apiserver-amd64:v1.9.0
      sizeBytes: 210373620
    - names:
      - gcr.io/google_containers/kube-apiserver-amd64@sha256:6c74eb9d3aed062d604a518f7e346ea92ee4466ede47671b756d780e3993e6f7
      - gcr.io/google_containers/kube-apiserver-amd64:v1.8.5
      sizeBytes: 194392646
    - names:
      - gcr.io/google_containers/kube-apiserver-amd64@sha256:f474819f3ebf18a064260e86fdca04f56a744db5c0d29741bc1bc461b6d5f223
      - gcr.io/google_containers/kube-apiserver-amd64:v1.8.4
      sizeBytes: 194371442
    - names:
      - gcr.io/google_containers/etcd-amd64@sha256:28cf78933de29fd26d7a879e51ebd39784cd98109568fd3da61b141257fb85a6
      - gcr.io/google_containers/etcd-amd64:3.1.10
      sizeBytes: 192743523
    - names:
      - gcr.io/google_containers/etcd-amd64@sha256:d83d3545e06fb035db8512e33bd44afb55dea007a3abd7b17742d3ac6d235940
      - gcr.io/google_containers/etcd-amd64:3.0.17
      sizeBytes: 168935516
    - names:
      - gcr.io/google_containers/kube-controller-manager-amd64@sha256:c65affb682b51f8137e52841732c90170d1b9f2ec32202be0be124dff3d3ba39
      - gcr.io/google_containers/kube-controller-manager-amd64:v1.9.0
      sizeBytes: 137750466
    - names:
      - gcr.io/google_containers/kube-controller-manager-amd64@sha256:29f08121733e1ff54f3f72dbb4638729d4d457b5e7304691d2b814dfaa3400fc
      - gcr.io/google_containers/kube-controller-manager-amd64:v1.8.5
      sizeBytes: 129336200
    - names:
      - gcr.io/google_containers/kube-controller-manager-amd64@sha256:8adbcd2de1b1ef752ce92c0602f99aa4bd86798c7b546e56c398e18f9f60c26b
      - gcr.io/google_containers/kube-controller-manager-amd64:v1.8.4
      sizeBytes: 129314896
    - names:
      - gcr.io/google_containers/kubernetes-dashboard-amd64@sha256:71a0de5c6a21cb0c2fbcad71a4fef47acd3e61cd78109822d35e1742f9d8140d
      - gcr.io/google_containers/kubernetes-dashboard-amd64:v1.8.0
      sizeBytes: 119155776
    - names:
      - gcr.io/google_containers/kube-proxy-amd64@sha256:3760355108b97ed0ff74d33fbec72764e1d7f8fbdb06ed0934117196819e3d00
      - gcr.io/google_containers/kube-proxy-amd64:v1.9.0
      sizeBytes: 109093516
    - names:
      - gcr.io/google_containers/kube-proxy-amd64@sha256:75e0c95306fa74a5ae2078aaf3e71513a787d7d84432b4d6c6c444150546bb4b
      - gcr.io/google_containers/kube-proxy-amd64:v1.8.5
      sizeBytes: 93210129
    - names:
      - gcr.io/google_containers/kube-proxy-amd64@sha256:8dce98802846d5219093f0313dcb8697a8a5d7cad647c3b71a816cf3a2b2eb2a
      - gcr.io/google_containers/kube-proxy-amd64:v1.8.4
      sizeBytes: 93197153
    - names:
      - gcr.io/google_containers/kube-scheduler-amd64@sha256:14f8f8dcf2e52e19b5db5ad8e84239d50fdb080a8b96677f9cf0ea944b40bec4
      - gcr.io/google_containers/kube-scheduler-amd64:v1.9.0
      sizeBytes: 62696260
    - names:
      - gcr.io/google_containers/kube-scheduler-amd64@sha256:72608b511275a1661b66f113cff09a0737b4d8e1405ad0ddb2e98c9cad0a8323
      - gcr.io/google_containers/kube-scheduler-amd64:v1.8.4
      sizeBytes: 54976083
    - names:
      - gcr.io/google_containers/kube-scheduler-amd64@sha256:03c0cc3642b75f873a878129f60a934120d6202914dfddb55ab98495c63039af
      - gcr.io/google_containers/kube-scheduler-amd64:v1.8.5
      sizeBytes: 54976083
    - names:
      - quay.io/coreos/flannel@sha256:056cf57fd3bbe7264c0be1a3b34ec2e289b33e51c70f332f4e88aa83970ad891
      - quay.io/coreos/flannel:v0.9.1-amd64
      sizeBytes: 51314612
    - names:
      - gcr.io/google_containers/k8s-dns-kube-dns-amd64@sha256:1a3fc069de481ae690188f6f1ba4664b5cc7760af37120f70c86505c79eea61d
      - gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.5
      sizeBytes: 49383083
    - names:
      - gcr.io/google_containers/k8s-dns-sidecar-amd64@sha256:9aab42bf6a2a068b797fe7d91a5d8d915b10dbbc3d6f2b10492848debfba6044
      - gcr.io/google_containers/k8s-dns-sidecar-amd64:1.14.5
      sizeBytes: 41814849
    - names:
      - gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64@sha256:46b933bb70270c8a02fa6b6f87d440f6f1fce1a5a2a719e164f83f7b109f7544
      - gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64:1.14.5
      sizeBytes: 41419289
    - names:
      - gcr.io/google_containers/pause-amd64@sha256:163ac025575b775d1c0f9bf0bdd0f086883171eb475b5068e7defa4ca9e76516
      - gcr.io/google_containers/pause-amd64:3.0
      sizeBytes: 746888
    nodeInfo:
      architecture: amd64
      bootID: 0f62dda3-a2a2-4306-b8e1-a907eefe628f
      containerRuntimeVersion: docker://1.12.6
      kernelVersion: 3.10.0-693.11.6.el7.x86_64
      kubeProxyVersion: v1.8.4
      kubeletVersion: v1.8.4
      machineID: b2ac3dffb1e94e46a095e82d6e157102
      operatingSystem: linux
      osImage: CentOS Linux 7 (Core)
      systemUUID: 420D0A29-1496-D50C-4C0E-B58AFFF5E912
- apiVersion: v1
  kind: Node
  metadata:
    annotations:
      flannel.alpha.coreos.com/backend-data: '{"VtepMAC":"c6:ed:11:e5:f1:55"}'
      flannel.alpha.coreos.com/backend-type: vxlan
      flannel.alpha.coreos.com/kube-subnet-manager: "true"
      flannel.alpha.coreos.com/public-ip: 10.1.1.2
      node.alpha.kubernetes.io/ttl: "0"
      volumes.kubernetes.io/controller-managed-attach-detach: "true"
    creationTimestamp: 2018-01-31T20:01:24Z
    labels:
      beta.kubernetes.io/arch: amd64
      beta.kubernetes.io/os: linux
      kubernetes.io/hostname: worker1.company.com
    name: worker1.company.com
    namespace: ""
    resourceVersion: "2355338"
    selfLink: /api/v1/nodes/worker1.company.com
    uid: 83f8f585-06c1-11e8-a925-0050568db8e3
  spec:
    externalID: worker1.company.com
    podCIDR: 10.244.1.0/24
  status:
    addresses:
    - address: 10.1.1.2
      type: InternalIP
    - address: worker1.company.com
      type: Hostname
    allocatable:
      cpu: "2"
      memory: 3761852Ki
      pods: "110"
    capacity:
      cpu: "2"
      memory: 3864252Ki
      pods: "110"
    conditions:
    - lastHeartbeatTime: 2018-02-05T09:21:17Z
      lastTransitionTime: 2018-01-31T20:01:24Z
      message: kubelet has sufficient disk space available
      reason: KubeletHasSufficientDisk
      status: "False"
      type: OutOfDisk
    - lastHeartbeatTime: 2018-02-05T09:21:17Z
      lastTransitionTime: 2018-01-31T21:26:15Z
      message: kubelet has sufficient memory available
      reason: KubeletHasSufficientMemory
      status: "False"
      type: MemoryPressure
    - lastHeartbeatTime: 2018-02-05T09:21:17Z
      lastTransitionTime: 2018-01-31T21:26:15Z
      message: kubelet has no disk pressure
      reason: KubeletHasNoDiskPressure
      status: "False"
      type: DiskPressure
    - lastHeartbeatTime: 2018-02-05T09:21:17Z
      lastTransitionTime: 2018-01-31T21:26:15Z
      message: kubelet is posting ready status
      reason: KubeletReady
      status: "True"
      type: Ready
    daemonEndpoints:
      kubeletEndpoint:
        Port: 10250
    images:
    - names:
      - docker.io/jupyterhub/k8s-singleuser-sample@sha256:7a1aa9992b0bec88ae37ebd37bd9e0f8bffe7845363ece5a8a402eedca375950
      - docker.io/jupyterhub/k8s-singleuser-sample:v0.6
      sizeBytes: 849956418
    - names:
      - docker.io/jupyterhub/k8s-singleuser-sample@sha256:31efa619c7440660dd34f2d560615003a2e5659cd5904a1a5b84e3fc4f018741
      - docker.io/jupyterhub/k8s-singleuser-sample:v0.5.0
      sizeBytes: 623721149
    - names:
      - docker.io/jupyterhub/k8s-hub@sha256:68634db503c07ccb22113ae2ad6c432297978acd829effb9100244fd153a96a2
      - docker.io/jupyterhub/k8s-hub:v0.5.0
      sizeBytes: 613269055
    - names:
      - quay.io/kubernetes-ingress-controller/nginx-ingress-controller@sha256:39cc6ce23e5bcdf8aa78bc28bbcfe0999e449bf99fe2e8d60984b417facc5cd4
      - quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.9.0
      sizeBytes: 189866039
    - names:
      - gcr.io/google_containers/kube-proxy-amd64@sha256:8dce98802846d5219093f0313dcb8697a8a5d7cad647c3b71a816cf3a2b2eb2a
      - gcr.io/google_containers/kube-proxy-amd64:v1.8.4
      sizeBytes: 93197153
    - names:
      - gcr.io/google_containers/heapster-amd64@sha256:b77cebeff2180d03e21cc9f9c6b69a0d9710caa9f6263e675eab7938019631ef
      - gcr.io/google_containers/heapster-amd64:v1.4.0
      sizeBytes: 73395475
    - names:
      - gcr.io/kubernetes-helm/tiller@sha256:df7f227fa722afc4931c912c1cad2c47856ec94f4d052ccceebcb16dd483dad8
      - gcr.io/kubernetes-helm/tiller:v2.7.2
      sizeBytes: 56572780
    - names:
      - docker.io/jupyterhub/configurable-http-proxy@sha256:c36cf3cc1c99f59348a8d6f5f64752df3eb4d88df93a11bc4e00acf23dbecfba
      - docker.io/jupyterhub/configurable-http-proxy:3.0.0
      sizeBytes: 56259118
    - names:
      - quay.io/coreos/flannel@sha256:056cf57fd3bbe7264c0be1a3b34ec2e289b33e51c70f332f4e88aa83970ad891
      - quay.io/coreos/flannel:v0.9.1-amd64
      sizeBytes: 51314612
    - names:
      - docker.io/traefik@sha256:a60f36880bbe8049cffb05f2f5d00d3068a327dfaebf85dab025c1d23b91f5bb
      - docker.io/traefik:v1.5.1
      sizeBytes: 49516718
    - names:
      - docker.io/yuvipanda/image-puller@sha256:3d40a8c98a2e0be4a40581dd26048a0e3a4ff2ed29a522b55c78430439978ba2
      - docker.io/yuvipanda/image-puller:v0.2
      sizeBytes: 6465997
    - names:
      - docker.io/jupyterhub/k8s-network-tools@sha256:271e7f7c9b4aa319788ca1bbce3428cad7b9da852a8c3af49178abcc2cc45b28
      - docker.io/jupyterhub/k8s-network-tools:v0.6
      sizeBytes: 5549803
    - names:
      - docker.io/jupyterhub/k8s-pre-puller@sha256:66edc6aad8055bcf77ea980c8c325b9e41456b1de1c64392e980ca954f75b2e1
      - docker.io/jupyterhub/k8s-pre-puller:v0.6
      sizeBytes: 4162080
    - names:
      - gcr.io/google_containers/pause-amd64@sha256:163ac025575b775d1c0f9bf0bdd0f086883171eb475b5068e7defa4ca9e76516
      - gcr.io/google_containers/pause@sha256:0d093c962a6c2dd8bb8727b661e2b5f13e9df884af9945b4cc7088d9350cd3ee
      - gcr.io/google_containers/pause-amd64:3.0
      - gcr.io/google_containers/pause:3.0
      sizeBytes: 746888
    nodeInfo:
      architecture: amd64
      bootID: 6812a258-6f41-415d-a9e2-b7bad8bcf0be
      containerRuntimeVersion: docker://1.12.6
      kernelVersion: 3.10.0-693.11.6.el7.x86_64
      kubeProxyVersion: v1.8.4
      kubeletVersion: v1.8.4
      machineID: 55bbf3b682fe4b9c862c73dd8f024b28
      operatingSystem: linux
      osImage: CentOS Linux 7 (Core)
      systemUUID: 420DD509-BE4A-D8B7-581F-BC0A5850A75C
- apiVersion: v1
  kind: Node
  metadata:
    annotations:
      flannel.alpha.coreos.com/backend-data: '{"VtepMAC":"5a:bc:e2:27:34:c8"}'
      flannel.alpha.coreos.com/backend-type: vxlan
      flannel.alpha.coreos.com/kube-subnet-manager: "true"
      flannel.alpha.coreos.com/public-ip: 10.1.1.3
      node.alpha.kubernetes.io/ttl: "0"
      volumes.kubernetes.io/controller-managed-attach-detach: "true"
    creationTimestamp: 2018-01-16T15:34:28Z
    labels:
      beta.kubernetes.io/arch: amd64
      beta.kubernetes.io/os: linux
      kubernetes.io/hostname: worker2.company.com
    name: worker2.company.com
    namespace: ""
    resourceVersion: "2355341"
    selfLink: /api/v1/nodes/worker2.company.com
    uid: bd5e524f-fad2-11e7-a5bb-0050568db8e3
  spec:
    externalID: worker2.company.com
    podCIDR: 10.244.2.0/24
  status:
    addresses:
    - address: 10.1.1.3
      type: InternalIP
    - address: worker2.company.com
      type: Hostname
    allocatable:
      cpu: "2"
      memory: 3761852Ki
      pods: "110"
    capacity:
      cpu: "2"
      memory: 3864252Ki
      pods: "110"
    conditions:
    - lastHeartbeatTime: 2018-02-05T09:21:17Z
      lastTransitionTime: 2018-01-16T15:34:28Z
      message: kubelet has sufficient disk space available
      reason: KubeletHasSufficientDisk
      status: "False"
      type: OutOfDisk
    - lastHeartbeatTime: 2018-02-05T09:21:17Z
      lastTransitionTime: 2018-02-01T08:42:40Z
      message: kubelet has sufficient memory available
      reason: KubeletHasSufficientMemory
      status: "False"
      type: MemoryPressure
    - lastHeartbeatTime: 2018-02-05T09:21:17Z
      lastTransitionTime: 2018-02-01T08:42:40Z
      message: kubelet has no disk pressure
      reason: KubeletHasNoDiskPressure
      status: "False"
      type: DiskPressure
    - lastHeartbeatTime: 2018-02-05T09:21:17Z
      lastTransitionTime: 2018-02-01T08:42:40Z
      message: kubelet is posting ready status
      reason: KubeletReady
      status: "True"
      type: Ready
    daemonEndpoints:
      kubeletEndpoint:
        Port: 10250
    images:
    - names:
      - docker.io/jupyterhub/k8s-singleuser-sample@sha256:7a1aa9992b0bec88ae37ebd37bd9e0f8bffe7845363ece5a8a402eedca375950
      - docker.io/jupyterhub/k8s-singleuser-sample:v0.6
      sizeBytes: 849956418
    - names:
      - docker.io/jupyterhub/k8s-singleuser-sample@sha256:31efa619c7440660dd34f2d560615003a2e5659cd5904a1a5b84e3fc4f018741
      - docker.io/jupyterhub/k8s-singleuser-sample:v0.5.0
      sizeBytes: 623721149
    - names:
      - docker.io/jupyterhub/k8s-hub@sha256:04dd7d25d348016ec009647399adcea2833fc53a66e66e0c288d333e7c3aabd8
      - docker.io/jupyterhub/k8s-hub:v0.6
      sizeBytes: 493573480
    - names:
      - quay.io/kubernetes-ingress-controller/nginx-ingress-controller@sha256:39cc6ce23e5bcdf8aa78bc28bbcfe0999e449bf99fe2e8d60984b417facc5cd4
      - quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.9.0
      sizeBytes: 189866039
    - names:
      - docker.io/yuvipanda/image-allnodes-puller@sha256:662189f5243ef5eba41bda71dd842f43ad7b53184b29deee0465676de0c2ad51
      - docker.io/yuvipanda/image-allnodes-puller:v0.8
      sizeBytes: 105852927
    - names:
      - gcr.io/google_containers/kube-proxy-amd64@sha256:8dce98802846d5219093f0313dcb8697a8a5d7cad647c3b71a816cf3a2b2eb2a
      - gcr.io/google_containers/kube-proxy-amd64:v1.8.4
      sizeBytes: 93197153
    - names:
      - docker.io/jupyterhub/configurable-http-proxy@sha256:c36cf3cc1c99f59348a8d6f5f64752df3eb4d88df93a11bc4e00acf23dbecfba
      - docker.io/jupyterhub/configurable-http-proxy:3.0.0
      sizeBytes: 56259118
    - names:
      - quay.io/coreos/flannel@sha256:056cf57fd3bbe7264c0be1a3b34ec2e289b33e51c70f332f4e88aa83970ad891
      - quay.io/coreos/flannel:v0.9.1-amd64
      sizeBytes: 51314612
    - names:
      - docker.io/traefik@sha256:a60f36880bbe8049cffb05f2f5d00d3068a327dfaebf85dab025c1d23b91f5bb
      - docker.io/traefik:v1.5.1
      sizeBytes: 49516718
    - names:
      - docker.io/yuvipanda/image-puller@sha256:3d40a8c98a2e0be4a40581dd26048a0e3a4ff2ed29a522b55c78430439978ba2
      - docker.io/yuvipanda/image-puller:v0.2
      sizeBytes: 6465997
    - names:
      - docker.io/jupyterhub/k8s-network-tools@sha256:271e7f7c9b4aa319788ca1bbce3428cad7b9da852a8c3af49178abcc2cc45b28
      - docker.io/jupyterhub/k8s-network-tools:v0.6
      sizeBytes: 5549803
    - names:
      - docker.io/jupyterhub/k8s-pre-puller@sha256:66edc6aad8055bcf77ea980c8c325b9e41456b1de1c64392e980ca954f75b2e1
      - docker.io/jupyterhub/k8s-pre-puller:v0.6
      sizeBytes: 4162080
    - names:
      - gcr.io/google_containers/pause-amd64@sha256:163ac025575b775d1c0f9bf0bdd0f086883171eb475b5068e7defa4ca9e76516
      - gcr.io/google_containers/pause@sha256:0d093c962a6c2dd8bb8727b661e2b5f13e9df884af9945b4cc7088d9350cd3ee
      - gcr.io/google_containers/pause-amd64:3.0
      - gcr.io/google_containers/pause:3.0
      sizeBytes: 746888
    nodeInfo:
      architecture: amd64
      bootID: 950eaf10-b4c1-416d-b9bf-a03758f1ef98
      containerRuntimeVersion: docker://1.12.6
      kernelVersion: 3.10.0-693.11.6.el7.x86_64
      kubeProxyVersion: v1.8.4
      kubeletVersion: v1.8.4
      machineID: c0ebcd72edfb468089a450845701851a
      operatingSystem: linux
      osImage: CentOS Linux 7 (Core)
      systemUUID: 420D9700-0029-D6A7-AA5D-CBF9123AECE9
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

```

**Comment by consideRatio** [comment]
The pre-puller will compare the needed images with the available images in the node, but only the first 50. Perhaps that is whats happened. A wprkaround until it is fixed could be to purge all images from the available nodes.

If you have multiple nodes, i think it is your master node that will block, as that has a lot of additional images on it, and have probably remained up for the longest of all nodes.

Quickfix: scale down to 0 nodes, scale back up to 1 node, and thereby starting fresh... I think.

I wrote about this here recently.
https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/476#issuecomment-363972524

**Comment by GerbenWelter** [comment]
Thank your for your comment and the write-up, but I think it doesn't apply in my case. I've seen the problem in my test cluster at work but also on my newly setup cluster at home. So it seems to be something in the puller code with regards to k8s.

**Comment by actionshrimp** [comment]
I'm seeing a similar issue using the docker for mac built-in kubernetes cluster. I've tried recreating the cluster, and wiping as many images as possible off my local docker but the problem seems to stick around. One of my images seems to be logged as pulled down correctly, and if I manually do `docker pull` both images are available on my machine. Here's the kubernetes event log:

```2018-02-16 17:09:28 +0000 GMT   2018-02-16 17:09:28 +0000 GMT   1         pre-pull-jupyterhub-imandra-1-1518800966-wd276.1513dd6f0c58f1b1     Pod         spec.initContainers{image-pull-singleuser}       Normal    Pulled                    kubelet, docker-for-desktop         Container image "eu.gcr.io/vocal-territory-126312/jupyterhub-imandra:2018-02-16-16-19" already present on machine
2018-02-16 17:09:28 +0000 GMT   2018-02-16 17:09:28 +0000 GMT   1         pre-pull-jupyterhub-imandra-1-1518800966-wd276.1513dd6f10e352ff     Pod         spec.initContainers{image-pull-singleuser}       Normal    Created                   kubelet, docker-for-desktop         Created container
2018-02-16 17:09:28 +0000 GMT   2018-02-16 17:09:28 +0000 GMT   1         pre-pull-jupyterhub-imandra-1-1518800966-wd276.1513dd6f1b53ebac     Pod         spec.initContainers{image-pull-singleuser}       Normal    Started                   kubelet, docker-for-desktop         Started container
2018-02-16 17:09:29 +0000 GMT   2018-02-16 17:09:29 +0000 GMT   1         pre-pull-jupyterhub-imandra-1-1518800966-wd276.1513dd6f587ea1a4     Pod         spec.initContainers{image-pull-metadata-block}   Normal    Pulled                    kubelet, docker-for-desktop         Container image "jupyterhub/k8s-network-tools:v0.6" already present on machine
2018-02-16 17:09:29 +0000 GMT   2018-02-16 17:09:29 +0000 GMT   1         pre-pull-jupyterhub-imandra-1-1518800966-wd276.1513dd6f5f81c573     Pod         spec.initContainers{image-pull-metadata-block}   Normal    Created                   kubelet, docker-for-desktop         Created container
2018-02-16 17:09:29 +0000 GMT   2018-02-16 17:09:29 +0000 GMT   1         pre-pull-jupyterhub-imandra-1-1518800966-wd276.1513dd6f68d29864     Pod         spec.initContainers{image-pull-metadata-block}   Normal    Started                   kubelet, docker-for-desktop         Started container
2018-02-16 17:09:30 +0000 GMT   2018-02-16 17:09:30 +0000 GMT   1         pre-pull-jupyterhub-imandra-1-1518800966-wd276.1513dd6f954a796c     Pod         spec.containers{pause}                           Normal    Pulled                    kubelet, docker-for-desktop         Container image "gcr.io/google_containers/pause:3.0" already present on machine
2018-02-16 17:09:30 +0000 GMT   2018-02-16 17:09:30 +0000 GMT   1         pre-pull-jupyterhub-imandra-1-1518800966-wd276.1513dd6f9a7d4edc     Pod         spec.containers{pause}                           Normal    Created                   kubelet, docker-for-desktop         Created container
2018-02-16 17:09:30 +0000 GMT   2018-02-16 17:09:30 +0000 GMT   1         pre-pull-jupyterhub-imandra-1-1518800966-wd276.1513dd6fa2ccd01f     Pod         spec.containers{pause}                           Normal    Started                   kubelet, docker-for-desktop         Started container
2018-02-16 17:09:28 +0000 GMT   2018-02-16 17:09:28 +0000 GMT   1         pre-pull-jupyterhub-imandra-1-1518800966.1513dd6ef72283b5           DaemonSet                                                    Normal    SuccessfulCreate          daemonset-controller                Created pod: pre-pull-jupyterhub-imandra-1-1518800966-wd276
2018-02-16 17:09:27 +0000 GMT   2018-02-16 17:09:27 +0000 GMT   1         pre-puller-1518800966-jupyterhub-imandra-1-nxttc.1513dd6ec057d0f7   Pod                                                          Normal    Scheduled                 default-scheduler                   Successfully assigned pre-puller-1518800966-jupyterhub-imandra-1-nxttc to docker-for-desktop
2018-02-16 17:09:27 +0000 GMT   2018-02-16 17:09:27 +0000 GMT   1         pre-puller-1518800966-jupyterhub-imandra-1-nxttc.1513dd6ecadf47f6   Pod                                                          Normal    SuccessfulMountVolume     kubelet, docker-for-desktop         MountVolume.SetUp succeeded for volume "pre-puller-1518800966-jupyterhub-imandra-1-token-stkbb"
2018-02-16 17:09:27 +0000 GMT   2018-02-16 17:09:27 +0000 GMT   1         pre-puller-1518800966-jupyterhub-imandra-1-nxttc.1513dd6ee72385ae   Pod         spec.containers{pre-puller}                      Normal    Pulled                    kubelet, docker-for-desktop         Container image "jupyterhub/k8s-pre-puller:v0.6" already present on machine
2018-02-16 17:09:27 +0000 GMT   2018-02-16 17:09:27 +0000 GMT   1         pre-puller-1518800966-jupyterhub-imandra-1-nxttc.1513dd6eed1d618f   Pod         spec.containers{pre-puller}                      Normal    Created                   kubelet, docker-for-desktop         Created container
2018-02-16 17:09:27 +0000 GMT   2018-02-16 17:09:27 +0000 GMT   1         pre-puller-1518800966-jupyterhub-imandra-1-nxttc.1513dd6ef568c981   Pod         spec.containers{pre-puller}                      Normal    Started                   kubelet, docker-for-desktop         Started container
2018-02-16 17:09:27 +0000 GMT   2018-02-16 17:09:27 +0000 GMT   1         pre-puller-1518800966-jupyterhub-imandra-1.1513dd6ebfe977a3         Job                                                          Normal    SuccessfulCreate          job-controller                      Created pod: pre-puller-1518800966-jupyterhub-imandra-1-nxttc
```

And the log from `pre-puller` outputs:
```2018/02/16 17:23:41 1 of 2 images pulled in node docker-for-desktop
2018/02/16 17:23:43 1 of 2 images pulled in node docker-for-desktop
2018/02/16 17:23:45 1 of 2 images pulled in node docker-for-desktop
```

Hope this is of some use, let me know if I can provide any other useful info :-)

**Comment by consideRatio** [comment]
REBOOTING INTO LINUX - COMMENT WILL BE UPDATED - TEMP STORAGE... actionshrimp GerbenWelter #502 might solve your problems, could you help me verify if that is the case?

```shell
# Current API
http://localhost:8080/api/v1/nodes
```
__response__
```json
{
  "kind": "PodList",
  "apiVersion": "v1",
  "metadata": {
    "selfLink": "/api/v1/pods",
    "resourceVersion": "5428524"
  },
  "items": [
    {
      ...
      "status": {
        "phase": "Running",
        ...
```


### Explanation
When you "read <X> of <Y> images pulled in node <nodename>" you are reading the output of a script that does the following.

It queries the k8s API on /api/v1/nodes which returns a NodeList containing Node items, each Node item will list up to 50 images it has. If you want to confirm that you have or not have the bug I had and if it will be solved by my PR. You could do the following.

**Comment by actionshrimp** [comment]
Thanks, I'll try and give it a go - is the easiest way to run `helm` on a version of this repo checked out on your branch?

**Comment by consideRatio** [comment]
@actionshrimp Hmm I have not thought about how you could utlize my code even before the PR is accepted. Images / Helm charts hmmm... so if you checkout my branch... ehhh wrapping my head around all the details here is tough... 

Step by step...
- I have not published any helm chart
- My built docker images are stored privately (doh - i'll make them public in the future)
- You can checkout my branch and use the build tool there with `./build.py --push --deploy --tag <sometag>` i think (not used that tool for a while).
- I have currently made a silly breaking change where i renamed prePuller to imagePuller in values, if you provide additional values files, make sure to update them.

But before you try my fix, could you diagnose your current situation by running commands in the comment above and providing me with the output?

Running to the store before it closes (22.00 soon here)

**Comment by actionshrimp** [comment]
Ah yep `kubectl get node docker-for-desktop -o json | grep "sizeBytes" | wc -l` gives me 50

I only have one node as it's my local cluster, and there was one instance of both the pre-pull pod controlled by the DS, and the pre-puller pod running (which was apparently stuck in a loop)

**Comment by GerbenWelter** [comment]
In my case on my single node test cluster at home `kubectl get node $(hostname -f) -o json | grep "sizeBytes" | wc -l` gave me 22 before starting the deployment of jupyterhub and 25 after. But then the pre-puller pod gets stuck in the `0 of 2 images pulled` loop.

**Comment by consideRatio** [comment]
@actionshrimp okay then you will risk getting stuck unless the images are within those 50 responses, they seem to be sorted by size, and the network-tooling thing is a very small size.

You could potentially temporarily workaround your issue by setting `singleuser.cloudMetadata.enabled: true` in values, and remove the need for the smaller networkTools image. I think that might be the image you don't find, as the results _might_ be sorted with descending size and this image has a very small size.

@GerbenWelter 
I do not know what fails, but if you want to debug it further, you could verify some steps.

1. Conclude what images are actually pulled by inspecting the daemonset pod spec directly (`kubectl describe daemonset <pre-puller-...>`), or the `values.yaml` indirectly (`singleuser.image.name` and `singleuser.networkTools.image.name`). The pulling is done by the daemonset pods, and the deaemonset is created by this helper, https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/76fcf7b208d19593ff5ce536a8319f6a3400e714/jupyterhub/templates/pre-puller/_helpers.yaml#L32-L46 
2. `kubectl proxy --port 8080` then `curl http://localhost:8080/api/v1/nodes` and inpect the images available, are they found? 
3. Perhaps you could end up figuring out whats wrong with the [main.go / node.go](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/tree/master/images/pre-puller) code in the prepuller image by now. If you inspect this code, I found it relevant to look at [the kubernetes API for NodeList](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#node-v1-core).

**Comment by consideRatio** [comment]
@GerbenWelter @actionshrimp the #502 PR updating the pre-puller logic got merged recently! I hope that it solves both of your issues, please let me know if you try it!

**Comment by eexe1** [comment]
using the release: jupyterhub-v0.7-360849f, in my case, the issue was solved.

**Comment by consideRatio** [comment]
### Summary
- PR #502 fixes issues with the image puller (not getting stuck waiting for images to be pulled, which are actually pulled, for example). The PR was merged after the v0.6 chart release.
- Helm 2.8.2 fixes other issues (regarding deletion of temporary hook-image-puller objects)

I'm closing this issue, please report back if the v0.7-dev releases + helm 2.8.2 still causes problems with the image-puller daemonsets or the pods they create. Note that only the continuous image puller daemonset and the pods created by it should remain after upgrades.

(@tracek - I got confused when we spoke thinking you utilized a release containing #502, but I was wrong.)

**Comment by EamonKeane** [comment]
Ran into this issue today, seemed to only be an issue for multiple node pools. Fun two hours.
